Data Structures/Trees/generic_tree.cpp

```diff
@@ -0,0 +1,318 @@
+//Name:Anuj Goyal
+//Email:anujg935@gmail.com
+
+//This program is to implement Generic Tree data structure
+//with following custom function:
+	// Level Wise Take Input Function
+  	//Level Wise Print Output Function
+  	//Counting Number Of Nodes Function
+  	//Sum Of Data Of all Nodes Function
+  	//Maximum Data Node Function
+  	//Count Leaf Node Function
+  	//Find Height of Tree Function
+  	//Print at Depth/Level Function
+  	//Pre Order Traversal
+  	//Post Order Traversal
+  	
+  	
+#include<iostream>
+#include<queue>
+
+using namespace std;
+
+//template for the generic datatype
+template <typename T>
+
+//GenericTreeNode class for generic tree 
+class GenericTreeNode{
+	public:
+		//T type data of node
+		T data;
+		//vector children of type GenericTreeNode to store the address of its children nodes 
+		vector<GenericTreeNode<T>*> children;
+		//constructor to initialize the data at starting
+		GenericTreeNode(T data){
+			this->data = data;
+		}
+		//destructor to delete all the tree
+		~GenericTreeNode(){
+			//this for loop is used for deleting the chicldrens of the root 
+			//as if i delete only the root then all the remaining childrens left undeleted
+			for(int i=0;i<children.size();i++){
+				delete children[i];
+			}
+		}
+		
+};
+
+//takeInput() function takes the input from the user level wise
+GenericTreeNode<int>* takeInput(){
+	//prompt the user to enter the starting root data
+	int rootData;
+	cout<<"Enter root data :"<<endl;
+	cin>>rootData;
+	
+	//created an object of class GenericTreeNode root with input of rootData
+	GenericTreeNode<int>* root = new GenericTreeNode<int>(rootData);

```

Comment:
@Anujg935
I think you should take input of datatype T otherwise there is no benefit of template as you have restricted your object internally to take input of type 'int'